tests/console/osc.pm

```diff
@@ -0,0 +1,81 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+# Author: Gustavo Figueira <gfigueira@suse.com>
+
+use base "consoletest";
+use testapi;
+use utils;
+use Utils::Backends 'use_ssh_serial_console';
+use version_utils qw(is_sle get_os_release is_tumbleweed);
+use strict;
+use warnings;
+
+sub repos_and_pkgs_sle {
+    my ($hub_version, $arch) = @_;
+    my $repo_version = $hub_version;
+    $repo_version =~ s/\./_SP/;
+    assert_script_run("SUSEConnect --product PackageHub/" . $hub_version . "/" . $arch);
+    zypper_call("ar http://download.suse.de/ibs/SUSE:/CA/SLE_" . $repo_version . "/ suseca");
+    zypper_call("in osc obs-service-source_validator ca-certificates-suse");
+}
+
+sub prepare_oscrc {
+    my ($oscrc_path, $IBSUSER, $SECRET) = @_;
+    assert_script_run("wget --quiet " . data_url('oscrc') . " -O " . $oscrc_path, timeout => 30);
+    assert_script_run("perl -pi -e 's/TAGIBSUSER/" . $IBSUSER . "/g' " . $oscrc_path);
+    assert_script_run("perl -pi -e 's/TAGSECRET/" . $SECRET . "/g' " . $oscrc_path);
+}
+
+sub co_and_build {
+    my ($codestream, $testpkg) = @_;
+    my $api_opensuse = "https://api.opensuse.org";
+
+    if (is_sle) {
+        assert_script_run("yes 1 | osc co " . $codestream . " " . $testpkg);
+    } else {
+        assert_script_run("yes 1 | osc -A " . $api_opensuse . " co " .  $codestream . " " . $testpkg);
+    }
+    assert_script_run("cd " . $codestream . "/" . $testpkg . "; yes 1 | osc build", timeout => 300);
+}
+
+sub run {
+    check_var("BACKEND", "ipmi") ? use_ssh_serial_console : select_console 'root-console';
+
+    my $IBSUSER = get_var("IBSUSER", "");
+    my $SECRET = get_var("SECRET", "");
+    my $arch = get_var("ARCH");
+    my ($sle_major_version, $sle_sp) = get_os_release;
+    my $testpkg = "perl-IO-String";
+    my $oscrc_path = "/root/.oscrc";
+    my $hub_version = $sle_major_version . "." . $sle_sp;
+    my $sle_cs = "SUSE:SLE-" . $sle_major_version . ":Update";
+
+    # 1) Repositories and packages needed
+    if (is_sle) {
+        repos_and_pkgs_sle($hub_version, $arch);
+    } else {
+        zypper_call("in osc");
+    }
+
+    # 2) Setting osc config file
+    prepare_oscrc($oscrc_path, $IBSUSER, $SECRET);
+
+    # 3) Checking-out and building locally
+    if (is_sle) {
+        co_and_build($sle_cs, $testpkg);
```

Comment:
does it make sense to build different packages on SLE vs. openSUSE?

Reply:
The package is always the same, what differs is the API and codestream. Does it make sense to use different APIs? In SLE we use the API where SLE is built upon. Does that answer your question?

Reply:
Well, I suggest to use good defaults and do what is possible to do without relying on internal instances. So as you use test variables how about using OBS (build.opensuse.org) by default and if corresponding test variables are set then have an additional SLE specific test?

Reply:
Also I suggest to think from customer perspective. A customer does not have access to IBS anyway but might want to use a SLE environment to use `osc` to build and develop OBS packages, right?

Reply:
The use case you propose (building a package locally without sourcing from an API) is a new one. The progress ticket explicitly states a "co" is needed.

Reply:
I am fine to source something from API, don't want to introduce a new use case. I only suggest to use the very same use case regardless of the underlying operating system.

Reply:
So you want the module to source from only one API? "I only suggest to use the very same use case regardless of the underlying operating system." The use case is the same IMHO, just the API is different because I think it makes sense to source from build.suse.de for SLE and build.opensuse.org for tumbleweed/leap. As a SLE users we are often consuming packages built in build.suse.de.

Reply:
> As a SLE users we are often consuming packages built in build.suse.de.

Yes, and a common suggestion for SUSE employees is to use an openSUSE desktop to do development, which can include both build.opensuse.org and build.suse.de

Based on my experience test reviewers that would see any test failures can get very confused if they see "the test works on openSUSE Tumbleweed but not on SLE" and it can take a long time to find out that the difference is that the test does not even do the same thing. This is why I suggest *by default* to have a simple test that simply does the same thing regardless of the environment. But I am open to receive feedback from other reviewers about this.